unsafe/src/main/java/org/apache/spark/unsafe/Platform.java

```diff
@@ -107,12 +107,27 @@ public static void freeMemory(long address) {
 
   public static void copyMemory(
     Object src, long srcOffset, Object dst, long dstOffset, long length) {
-    while (length > 0) {
-      long size = Math.min(length, UNSAFE_COPY_THRESHOLD);
-      _UNSAFE.copyMemory(src, srcOffset, dst, dstOffset, size);
-      length -= size;
-      srcOffset += size;
-      dstOffset += size;
+    // Check if dstOffset is before or after srcOffset to determine if we should copy
+    // forward or backwards. This is necessary in case src and dst overlap.
+    if (dstOffset < srcOffset) {
```

Comment:
In JVM, two objects (or memory block) can't overlap each other (assuming the two offsets are in  valid range). The only case that src and dst could overlap is they are exactly the same value (pointer), so we can only compare the offsets.

After this patch, we can use copyMemory() without worrying about the src and dst are overlapped or not, it always works.